music/src/core/player.ts

```diff
@@ -328,9 +328,89 @@ export class SoundFontPlayer extends BasePlayer {
 }
 
 /**
- * A `NoteSequence` player based on Tone.js that includes a click track.
+ * A base class for providing arbitrary callbacks for each note played.
  */
-export class PlayerWithClick extends Player {
+export class BasePlayerCallback {
+
+  /* The function should be overridden, it will be called for each time/note
+   * pair in a sequence being played.
+   */
+  run(t: number, n: NoteSequence.INote) {
+    // Do nothing.
+  }
+
+    /* The function should be overridden, it will be called when a sequence is
+     * stopped.
+     */
+  stop() {
+    // Do nothing.
+  }
+}
+
+/**
+ * A `NoteSequence` player based on Tone.js that allows arbitrary callbacks
+ * when playing each note.
+ */
+export class PlayerWithCallback extends Player {
```

Comment:
It seems like it would be better/easier just to add the callback as an option to Player.

Reply:
the way it's coded up i would have to add the constructor options to BasePlayer, since i modify the functionality of `start()`. just want to confirm you're ok with that before proceeding.

Reply:
Yes, that's fine as long as it works for all of the child classes.

Reply:
moved all click and callback logic into `BasePlayer`. made code a lot cleaner, and now all can benefit from this! :)